hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableConfig.java

```diff
@@ -781,7 +782,7 @@ public static Triple<RecordMergeMode, String, String> inferCorrectMergingBehavio
               + "strategy ID (%s).", payloadClassName, recordMergeStrategyId));
       // TODO(HUDI-8925): once payload class name is not required, remove the check on
       //  modeBasedOnStrategyId
-      if (modeBasedOnStrategyId != CUSTOM && modeBasedOnPayload != null && modeBasedOnStrategyId != null) {
+      if (getTableVersion(config).greaterThanOrEquals(HoodieTableVersion.EIGHT) && modeBasedOnStrategyId != CUSTOM && modeBasedOnPayload != null && modeBasedOnStrategyId != null) {
```

Comment:
this is the main fix i.e. do the validation only for table version 8 or greater

Reply:
Looks good to me.